### DISCUSSION TITLE
new 7 colors display

### DISCUSSION THREAD

**Discussion by Pouletgrill** [txoof/PaperPi#77] [General]
Hi, i am currently testing paperPi witha 2.13inch Black  and Red Epaper but was unable to use red on my slideshow, I i read correctly it seams to not be supporte yet but i was wondering if you were planning to support it soon ?
My real goal is to By the new 7 color 7.5Inch epaper display from Waveshare and use PaperPi. do you plan to support it in the future ?

I tried to edit the code myself to make the red work but did not succed for now.

**Comment by aaronr8684** [comment]
I started bringing in grey and color support a while ago and got caught up in other things. I'm right in the middle of another project, but could take a look when I wrap up. I've been meaning to pick this back up again.

**Reply by Pouletgrill** [reply]
If you manage to make it work with the epd_7in3f, i would be able to test it for you when i will receive it.
my order is in the mail 
https://www.waveshare.com/wiki/7.3inch_e-Paper_HAT_(F)_Manual#Python

**Reply by aaronr8684** [reply]
Sure thing. I have the 5.65f and 7.5c (red) and a bunch of other smaller ones, but they've really expanded their line (including yours) so it will be good to test as many as we can.

**Reply by txoof** [reply]
@aaronr8684 

I ended up rolling back the OmniEPD changes due to lack of grayscale support. I hope that doesn't throw out your work out too far. 

@Pouletgrill 
I would love to get color working. It will take some changes to the EPDLib at the backend to make that happen. All the image processing is done in the EPDLib module; that's where the color handling needs to happen eventually.

**Is there a good value multi-color screen you would recommend? ** I'd be interested in fiddling with a color screen.

I took a quick browse through the [Screen class](https://github.com/txoof/epdlib/blob/master/epdlib/Screen.py) and all the clear and update functions will need to be rewritten to handle multiple colors and decide which write functions to push at the screens.

**Reply by aaronr8684** [reply]
> I ended up rolling back the OmniEPD changes due to lack of grayscale support. I hope that doesn't throw out your work out too far.

All good. It's been so long since I looked at the code, I'd want a fresh take anyway. When I start playing around with it, I'll let you know if I need anything

**Reply by txoof** [reply]
Wow. It looks like wavehsare has cleaned up their methods a bit since last I looked. It might be as simple as calling the write/clear functions just like the black and white screens.

Maybe I'll pick up a 4.01" color and give it a go.

**Reply by aaronr8684** [reply]
Yea, they've also added a bunch of yellow/red/black screens which is new from the last time I looked. I might have to grab a few new ones too 😄

**Comment by txoof** [comment]
Step 1: Buy a 7 Color Display
Step 2: Test it
Step 3: ???
Step 4: Profit!?

I wasn't expecting super fast refresh, but wow. The 7 Color refresh time  is something else.

![PXL_20230205_184550605](https://user-images.githubusercontent.com/14262505/216838751-fd61769d-3f5e-4393-a9f3-cfbe542aa6e6.jpg)

**Reply by aaronr8684** [reply]
Is the refresh still upwards of 30 seconds? I wonder if the new screens are any faster?

**Reply by txoof** [reply]
Yeah, it's pretty slow, but the color pallet is reasonable.

**Comment by txoof** [comment]
It looks like it will be pretty simple to update epdlib to work with 7-color screens. At the moment all that's required is overriding the default `Screen.mode` attribute that defaults to `'1'` with `'RGB'` for non HD screens and then feed it a color image:
![PXL_20230205_195218539 MP](https://user-images.githubusercontent.com/14262505/216842477-88f3c49e-a162-45eb-b840-f75095c1c976.jpg)

We'll just need a way to pick the 7-color screens out reliably and rewrite the logic around the mode setting.

**Comment by txoof** [comment]
@Pouletgrill 

I have a proof of concept that basically works in `epdlib`. I still need to validate it and clean up some messes, but that's one step closer to adding support to PaperPi. 

Once I've sorted out a few things, I'll start working on implementing the color features in PaperPi. Once that's done, we'll need to update a few plugins to take advantage of the color. Right now everything is stuck in 1 or 8 bit gray mode.

![PXL_20230207_182950013 MP](https://user-images.githubusercontent.com/14262505/217334340-4f17cfdc-35e2-481e-a88b-f82a18803be4.jpg)

**Reply by Pouletgrill** [reply]
it's crazy, it's barely a week since I discovered your project.
I took a chance to write to you without knowing how much it would restart the project to the next level (color Waw!). 
very efficient guys, good job!

**Reply by txoof** [reply]
@Pouletgrill haha 

You caught me at a good time. I'm between other projects and have some free time for something new. 

I'll keep plugging away on this and see if I can't get something usable in the next week or so.

**Comment by txoof** [comment]
@aaronr8684 
I've got a new version of epdlib (0.6) in progress over on the [development](https://github.com/txoof/epdlib/tree/development) branch. It support RGB; it's also now possible to update the properties of layout blocks on the fly without recalculating everything.

I want to clean up [this](https://github.com/txoof/epdlib/blob/ddb98051107a1d07b4a460b190727d8c966f01f7/epdlib/Layout.py#L165-L166) deprecation warning and then I'll start testing it against PaperPi. I have a feeling that some of the older plugins are going to freak out when I make that change.

The next step after that is to write a plugin/layout that can actually take advantage of the screen!

**Reply by aaronr8684** [reply]
Let me know if you want me to test on a few screens.

**Reply by txoof** [reply]
@aaronr8684 
Would you mind testing this out using the development branch on whatever screens you have laying around? This works great on a 5.65F (7 color) screen. I haven't had a chance to test it on a 1 bit screen.

```python
from Screen import Screen
from Layout import Layout
from PIL import Image
from time import sleep


s = Screen()
s.epd = 'epd5in65f'

s.resolution

# layout that uses RGB -- not quite sure what this will do on non-color screens
l = Layout(resolution=s.resolution, mode='RGB')
l.layout = {
    'image' : {
        'type': 'ImageBlock',
        'image': True,
        'padding': 10,
        'width': 1,
        'height': 0.9,
        'abs_coordinates': (0, 0),
        'hcenter': True,
        'vcenter': True,
        'relative': False,
        'mode': 'RGB',
        'bkground': 'white'
    },
    'text': {
        'type': 'TextBlock',
        'image': None,
        'max_lines': 3,
        'bkground': 'Yellow',
        'fill': 'red',
        'width': 1,
        'height': .1,
        'abs_coordinates': (0, None),
        'relative': ['text', 'image'],
        'font': '../fonts/Open_Sans/OpenSans-Regular.ttf',
        'mode': 'RGB',
        'vcenter': True,
        'hcenter': True,
        
        
    }
}

# standard color w/ standard dithering
i = Image.open('../images/hubble.jpg')
l.update_contents({'image': i, 'text': 'Jackdaws love my big sphinx of quartz!'})

s.writeEPD(l.concat())

sleep(5)

# update the text block properties
l.update_block_props('text', props={'fill': 'blue', 'max_lines': 1})

# reduce the palette with no dithering -- probably looks awful
i_reduced = s.reduce_palette(i, palette=s.colors2pallette(), dither=False)


l.update_contents({'image': i_reduced, 'text': 'Jackdaws love my big sphinx of quartz!'})

s.writeEPD(l.concat())
sleep(5)

# switch the image block mode to 1 bit
l.update_block_props('image', props={'mode': '1'})
# update the properties of the text block
l.update_block_props('text', props={'bkground': 0, 'fill': 'White'})

# write the standard color image in 1 bit mode
l.update_contents({'image': i, 'text': 'The quick brown fox jumps over the lazy dog'})

s.writeEPD(l.concat())

```

**Reply by txoof** [reply]
Something about the `hubble.jpg` was breaking my 5.83 screen and causing terrible output. I have no idea what the problem is, but you might want to switch it up with the `portrait-pilot` image if you get weird results.

**Reply by aaronr8684** [reply]
@txoof Here's my results

All of the screens were verified working before running your code via the example scripts provided by waveshare's library

- 5in65f (7-color) - PASSED
- 2in13_V3 (b/w) - PASSED (although the clearEPD that I added at the end of the code above failed)
- 3in7 (greyscale) - FAILED - unsupported (this was expected)
- 4in2b_V2 (yellow) - FAILED - see below
- 7in5b_V2 (red) - FAILED - see below

---

Both the 4.2 and 7.5 screens failed with the same error:
```python
Traceback (most recent call last):
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 635, in _full_writeEPD_non_hd
    self.epd.display(image_buffer, self.buffer_no_image)
  File "/home/pi/epd/epdlib/epdlib/waveshare_epd/epd4in2b_V2.py", line 127, in display
    self.send_data2(imagered)
  File "/home/pi/epd/epdlib/epdlib/waveshare_epd/epd4in2b_V2.py", line 73, in send_data2
    epdconfig.spi_writebyte2(data)
  File "/home/pi/epd/epdlib/epdlib/waveshare_epd/epdconfig.py", line 65, in spi_writebyte2
    self.SPI.writebytes2(data)
TypeError: Empty argument list.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/epd/epdlib/epdlib/test.py", line 51, in <module>
    s.writeEPD(l.concat())
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 273, in wrapper
    func(*args, **kwargs)
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 602, in writeEPD
    write_function(image)
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 638, in _full_writeEPD_non_hd
    raise ScreenError(f'failed to write image to display: {e}')
Screen.ScreenError: failed to write image to display: Empty argument list.
```

---

Here's the failure from the 2.13 clear attempt
```python
Traceback (most recent call last):
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 563, in _clearEPD_non_hd
    self.epd.Clear(**self.clear_args)
TypeError: Clear() missing 1 required positional argument: 'color'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/epd/epdlib/epdlib/test.py", line 79, in <module>
    s.clearEPD()
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 273, in wrapper
    func(*args, **kwargs)
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 547, in clearEPD
    return clear_function()
  File "/home/pi/epd/epdlib/epdlib/Screen.py", line 566, in _clearEPD_non_hd
    raise ScreenError(f'failed to clear screen: {e}')
Screen.ScreenError: failed to clear screen: Clear() missing 1 required positional argument: 'color'
```

**Reply by aaronr8684** [reply]
I wasn't sure if you wanted the result of the three tests for each screen too or just overall result. Let me know

**Reply by txoof** [reply]
Thanks for testing. Interesting errors. I'll have to dig into those.

I didn't think I changed much in the write/clear. Clearly I've done something major.

**Reply by txoof** [reply]
@aaronr8684 
Wow. You unearthed a spectacularly awesome bug that somehow has slipped through for a long time.

At least the `Screen.ScreenError: failed to write image to display: Empty argument list.` is due to a missing blank image. Setting `s.rotation = 0` fixes this as the rotation setter creates the blank image and stores it in `self.buffer_no_image`

When my brain is less toasted, I'll work on fixing that bug. It will take some rethinking of the whole init process, but shouldn't be too hard.

I'm not sure about the clear error yet.

Thanks again for the testing.

**Reply by txoof** [reply]
@aaronr8684  out of curiosity, do you hot-plug your screens when you are testing?

I was in the habit of swapping screens with the Pi powered up and never had a problem. Then, one day I managed to toast the driver board for my HD screen and spent months trying to buy a new driver from China. It was awesome.

**Reply by aaronr8684** [reply]
> @aaronr8684 out of curiosity, do you hot-plug your screens when you are testing?
> 
> I was in the habit of swapping screens with the Pi powered up and never had a problem. Then, one day I managed to toast the driver board for my HD screen and spent months trying to buy a new driver from China. It was awesome.

Depends on how lazy i'm feeling. I try to power down, but when i'm doing a quick test across all of my screens, sometimes i don't. That's good to know and I wonder if the HD screens are more sensitive to hot-swap. I know it's not technically supported and I definitely make sure the display is in sleep mode before i do. 

> I didn't think I changed much in the write/clear. Clearly I've done something major.

I also did some digging on the clear error and figured out the "if color" check wasn't detecting the parameter correctly (and probably unrelated to your changes). I opened a PR in EPDLIB with the fix (in the ipynb, let me know if you want me to copy the changes in the py as well). This should fix what I assume was a bug for the 1in54, 2in13, and 2in9 screens. One thing to note is that the clear color for the 2in7b is actually 0x00 instead of 0xFF (like the rest). I haven't updated my screen list in about a year, so this info might be out of date.

**Reply by txoof** [reply]
Thanks for the PR. I'll merge that in by hand. My git-fu is too weak to handle that on top of the work I'm doing for the other bug you found.

the `Clear()` function is a bit of free-for-all, but at least everything except for the 3in7 is standard at 0xFF. I've verified this in the [current repo](https://github.com/waveshare/e-Paper/tree/master/RaspberryPi_JetsonNano/python/examples). Fortunately everything without the default set, we can push `0xFF` at them and call it a day (see below).

I think the way to deal with this is to inspect the signature of the `Clear()` functions and if the color parameter is empty, assign it the default of 0xFF.

```
$ grep "Clear.*color.*" *                                            
epd1in54.py:    def Clear(self, color):
epd1in54_V2.py:    def Clear(self, color):
epd1in64g.py:    def Clear(self, color=0x55):
epd2in13.py:    def Clear(self, color):
epd2in13_V2.py:    def Clear(self, color):
epd2in13_V3.py:    def Clear(self, color):
epd2in36g.py:    def Clear(self, color=0x55):
epd2in7b.py:    def Clear(self, color=0x00):
epd2in7.py:    def Clear(self, color=0xFF):
epd2in9.py:    def Clear(self, color):
epd2in9_V2.py:    def Clear(self, color):
epd3in0g.py:    def Clear(self, color=0x55):
epd3in7.py:    def Clear(self, color, mode):
epd4in37g.py:    def Clear(self, color=0x55):
epd7in3f.py:    def Clear(self, color=0x11):
epd7in3g.py:    def Clear(self, color=0x55):
```

For my own reference:
* `1in54 = 0xFF`
* `1in54_V2 = 0xFF`
* `2in13 = 0xFF`
* `2in13_V2 = 0xFF`
* `2in13_v3 = 0xFF`
* `2in7 = 0xFF`
* `2in9_V2 = 0xFF`
* `3in7 = 0xFF, 0`

**Reply by aaronr8684** [reply]
> I think the way to deal with this is to inspect the signature of the `Clear()` functions and if the color parameter is empty, assign it the default of 0xFF.

I think this is definitely the best approach. That way, you're not overwriting the other defaults when it's not needed. I'll poke around if I have time and share what I find. 

I also opened a PR with waveshare to get the three 2in13 libraries to set a default for color, but that could take months based on other PRs I've seen.

**Reply by txoof** [reply]
here's what I'm tinkering with -- I edited this to fix a missing variable 
```Python
s = Screen()
# s.epd = 'epd4in2'
# s.epd = 'epd2in7'
s.epd = 'epd2in9_V2'
# s.epd = 'epd1in64g'
# s.epd = 'epd2in13_V3'
# s.epd = 'epd7in3f'
# s.epd = 'epd2in13'
# s.epd = 'epd3in7'

try:
    clear_sig = inspect.signature(s.epd.Clear)
except AttributeError as e:
    raise ScreenError(f'epd has an unsupported `EPD.Clear()` function and is not usable with this module ')

color_default = clear_sig.parameters.get('color', False)
    
if color_default:
    logging.debug(f'Clear function has color parameter')
    if color_default.default is color_default.empty:
        color_param = 0xFF
        logging.debug(f'Clear(color) parameter has no default using: {color_param}')
```

**Reply by txoof** [reply]
I also just discovered that some of the libraries don't have standard `EPD_HEIGHT` and `EPD_WIDTH` values. Sometimes it's written as `EPD_width`. 
¯\_(ツ)_/¯ FFS.

**Reply by txoof** [reply]
> I also just discovered that some of the libraries don't have standard EPD_HEIGHT and EPD_WIDTH values. Sometimes it's written as EPD_width.
epd3in52.py has `EPD_width` and `EPD_height` 

I wonder what other incompatibilities are lurking out there :)

**Reply by aaronr8684** [reply]
> here's what I'm tinkering with
> 
> ```python
> s = Screen()
> # s.epd = 'epd4in2'
> # s.epd = 'epd2in7'
> # s.epd = 'epd2in9_V2'
> # s.epd = 'epd2in13_V3'
> # s.epd = 'epd7in3f'
> # s.epd = 'epd2in13'
> s.epd = 'epd3in7'

This is similar to what mine looks like 😄 

---

> try:
>     clear_sig = inspect.signature(s.epd.Clear)
> except AttributeError as e:
>     raise ScreenError(f'epd has an unsupported `EPD.Clear()` function and is not usable with this module ')
> 
> color_param = None
> if color_default:
>     logging.debug(f'Clear function has color parameter')
>     if color_default.default is color_default.empty:
>         color_param = 0xFF
>         logging.debug(f'Clear(color) parameter has no default using: {color_param}')
> ```

What I came up with is changing this section in `Screen.py`:

```python
if 'color' in clear_args_spec:
    clear_args['color'] = 0xFF
```

to something like this:

```python
if 'color' in clear_args_spec.args and clear_args_spec.defaults is None:
    clear_args['color'] = 0xFF
```

This obviously isn't checking if the default set belongs to color, so that could be a little more robust, but it does work.

EDIT: I read more of your solution and it seems like a better solution once it's implemented on the Screen.py side.

**Reply by aaronr8684** [reply]
> I wonder what other incompatibilities are lurking out there :)

Yea the libraries are a mess. I know that was a huge advantage of the Omni_EPD and I wish I had more time to implement it 😐

**Comment by txoof** [comment]
> I also opened a PR with waveshare to get the three 2in13 libraries to set a default for color, but that could take months based on other PRs I've seen.

I went ahead and patched up all the inconsistent modules with the exception of the `epd3in7.py` module. I didn't have the patience to figure out what would be "correct" for the `mode` value.

**Reply by txoof** [reply]
https://github.com/waveshare/e-Paper/pull/280

**Reply by aaronr8684** [reply]
Nice, I went ahead and closed mine since you covered all my changes

**Reply by aaronr8684** [reply]
So from what I remember, the 3in7 has two issues. First, the mode=0 (4Gray) is a good default. Second, the "full update" method isn't `display(image)` but `display_4Gray(image)`

I had overloaded the functions to get it working with the old version of paperpi, but it definitely isn't a clean method. 

The main issue is that the 1Gray method uses getbuffer and the 4Gray method uses getbuffer_4Gray so you can't just create a display() in the driver for a full update because it won't work with the 1Gray buffer method.

**Reply by txoof** [reply]
uuug. That's fugly. I'm ok just leaving that screen as unsupported. Maybe they'll merge in the PR and make the rest somewhat "standard"

**Reply by txoof** [reply]
> I went ahead and patched up all the inconsistent modules with the exception of the epd3in7.py module. I didn't have the patience to figure out what would be "correct" for the mode value.

[Wow! PR accepted!](https://github.com/waveshare/e-Paper/pull/280#event-8533736232). That should make things way easier to handle! Only the 3in7 is unsupported now.

**Comment by txoof** [comment]
@Pouletgrill and @aaronr8684 

Good news! 

The updates to epdlib pretty much drop right in to PaperPi and 'just work'. There will need to be a few modifications to the loading the configuration and all the plugins and layouts will need to be updated to make use of 'RGB' mode, but check out the word clock and slideshow plugins below.

![image](https://user-images.githubusercontent.com/14262505/219148786-436d9082-7183-4516-992a-4d126bfd6ea8.png)
![image](https://user-images.githubusercontent.com/14262505/219149851-8e07037d-ca69-444b-83f4-216dea471045.png)
![PXL_20230215_205102855 MP](https://user-images.githubusercontent.com/14262505/219155320-5820b19a-125b-4ccf-9548-7a3cf0250260.jpg)

**Reply by txoof** [reply]
@aaronr8684 
I've pushed an updated version of epdlib to the development branch if you want to have a go with it against some more screens. Sometime in the next few days I'll see if I can't push a development branch of PaperPi out with the updated logic and configuration.

**Reply by aaronr8684** [reply]
5in65f (7-color) - FAILED (I think this is HW related...the library's example test doesn't work either. Hopefully I didn't fry it 😦)
2in13_V3 (b/w) - PASSED 
4in2b_V2 (yellow) - PASSED (bw only)
7in5b_V2 (red) - PASSED (bw only)

**Reply by aaronr8684** [reply]
Ok so I hooked up the screen to another control board and it's working fine. I think the integrated control board is just sitting at always high on the BUSY pin which causes the scripts to not wait for the screen to actually write the image.

**Reply by txoof** [reply]
I hope your board/screen is ok!

I guess I'll stick with no hot-plugging. 

Thanks for the testing.

**Comment by Pouletgrill** [comment]
Hello Guys, last week while testing my screen i found a bug with the rendering on picture that made it more yellow and fadded. 
I told waveshare and they just posted yesterday on their github the correction for my screen epd7in3f.py. 
I just wanted to make sur that you were working on up to date code before thinking that some fadding might be because of you while in this case is not. 
i let here the link for the commit from yesterday about my screen : https://github.com/waveshare/e-Paper/commit/8be47b27f1a6808fd82ea9ceeac04c172e4ee9a8

Cheers!

**Reply by txoof** [reply]
I just pulled in the latest version from Waveshare. The changes in the code you highlighted shouldn't have any impact on epdlib at all. 

Thanks for the heads up.

**Comment by txoof** [comment]
@Pouletgrill You can track progress here: https://github.com/txoof/PaperPi/issues/79

Sometime next week I hope to have a version that's ready to test.

**Comment by txoof** [comment]
@Pouletgrill and @aaronr8684 

I have a **very** early version ready for testing on the I#79_color_display branch. I've tested it against all the working screens I have and it seems OK. I don't have a working 2-color screen at the moment so if you could test against that, it would be really helpful!

If you all can try this out on the color screens you have that would be fantastic. 
This will automatically install from the development branch: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/txoof/PaperPi/I%2379_color_display/install/remote_install_development.sh)"`

You will need to add the following things to your global or user `paperpi.ini`

```ini
[main]
# if possible, use RGB colors; has no impact on 1, 2, 3 or grayscale screens
color = True
```

I've added RGB support for the following plugins:

- met_no
- word_clock
- slideshow
- reddit_quote
- lms_client
- librespot_client (I haven't had a chance to test this properly yet)

The following plugins will automatically display colored content without any additional configuration when using a 7-color screen:

- lms_client (album art)
- librespot_client (album art)
- slideshow (images)
- met_no (forecast images)

The following plugins support colored text/backgrounds, but you will need to add config options to each one:

- met_no
- reddit_quote
- word_clock

Add these configuration options to each of the plugins above to get colored text/background:

```ini
# choose from RED, ORANGE, YELLOW, GREEN, BLUE, WHITE, BLACK or random
text_color = RED
bkground = random
```

EXAMPLE:

```
[Plugin: Word Clock]
layout = layout
plugin = word_clock
refresh_rate = 200
min_display_time = 90
max_priority = 2
# Additional color configuration options here
text_color = random
bkground = BLACK
```

**Reply by aaronr8684** [reply]
Yeah I should be able to get to this in the next day or two

**Comment by txoof** [comment]
@Pouletgrill and @aaronr8684 

There's a working version with an updated ini available here: /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/txoof/PaperPi/I%2379_color_display/install/remote_install_development.sh)"

If you want to enable color for the word_clock or reddit plugins, you'll need to uncomment the color lines in the `paperpi.ini`, but the rest should just work as expected with a 7-color display.

Your experiences would be much appreciated.

**Comment by txoof** [comment]
@Pouletgrill and @aaronr8684 
Well, it's done!

If you want to have a play hit the standard installer and give it a whirl.

Your feedback is welcome! I'm going to go ahead and close this.

**Reply by aaronr8684** [reply]
I am unavailable this week, but should be able to play around with it next week
